contracts/src/token/erc20/extensions/flash_mint.rs

```diff
@@ -0,0 +1,447 @@
+        receiver: Address,
+        token: Address,
+        amount: U256,
+        data: Bytes,
+        erc20: &mut Erc20,
+    ) -> Result<bool, Self::Error>;
+}
+
+impl IErc3156FlashLender for Erc20FlashMint {
+    type Error = Error;
+
+    fn max_flash_loan(&self, token: Address, erc20: &Erc20) -> U256 {
+        if token == contract::address() {
+            return U256::MAX - erc20.total_supply();
+        }
+        U256::MIN
+    }
+
+    fn flash_fee(
+        &self,
+        token: Address,
+        _amount: U256,
+    ) -> Result<U256, Self::Error> {
+        if token != contract::address() {
+            return Err(Error::UnsupportedToken(ERC3156UnsupportedToken {
+                token,
+            }));
+        }
+        Ok(self.flash_fee_amount.get())
+    }
+
+    // This function can reenter, but it doesn't pose a risk because it always
+    // preserves the property that the amount minted at the beginning is always
+    // recovered and burned at the end, or else the entire function will revert.
+    fn flash_loan(
+        &mut self,
+        receiver: Address,
+        token: Address,
+        amount: U256,
+        data: Bytes,
+        erc20: &mut Erc20,
+    ) -> Result<bool, Self::Error> {
+        let max_loan = self.max_flash_loan(token, erc20);
+        if amount > max_loan {
+            return Err(Error::ExceededMaxLoan(ERC3156ExceededMaxLoan {
+                max_loan,
+            }));
+        }
+
+        let fee = self.flash_fee(token, amount)?;
+        if !Address::has_code(&receiver) {
+            return Err(Error::InvalidReceiver(ERC3156InvalidReceiver {
+                receiver,
+            }));
+        }
+        erc20._mint(receiver, amount)?;
+        let loan_receiver = IERC3156FlashBorrower::new(receiver);
+        let loan_return = loan_receiver
+            .on_flash_loan(
+                Call::new_in(self),
+                msg::sender(),
+                token,
+                amount,
+                fee,
+                data.to_vec().into(),
+            )
+            .map_err(|_| {
+                Error::InvalidReceiver(ERC3156InvalidReceiver { receiver })
+            })?;
+        if loan_return != BORROWER_CALLBACK_VALUE {
+            return Err(Error::InvalidReceiver(ERC3156InvalidReceiver {
+                receiver,
+            }));
+        }
+
+        let allowance = amount
+            .checked_add(fee)
+            .expect("allowance should not exceed `U256::MAX`");
+        erc20._spend_allowance(receiver, contract::address(), allowance)?;
+
+        let flash_fee_receiver = self.flash_fee_receiver_address.get();
+
+        if fee.is_zero() || flash_fee_receiver.is_zero() {
+            erc20._burn(receiver, allowance)?;
+        } else {
+            erc20._burn(receiver, amount)?;
+            erc20._transfer(receiver, flash_fee_receiver, fee)?;
+        }
+
+        Ok(true)
+    }
+}
+
+// TODO: unignore all tests once it's possible to mock contract address.
```

Comment:
Add GH issue for it.

Reply:
https://github.com/OpenZeppelin/rust-contracts-stylus/issues/470